Role List Table component

jerabekjiri opened the pull request:
No-Issue

Adding prototype component for basic Role List Table component that is identical on new/changed `RBAC` pages.

I decided to use a functional component and use hooks (`useState`). I've seen it's used for `collection signing` (@himdel is that okay? I can revert it to class component).

Use the component in code:
```
<RoleListTable
    roles={roles}
    dropdownActions={[
        <DropdownItem
            key='edit-role'
            onClick={(role) => {
                // edit role action
            }}
        >
            {t`Edit Role`}
        </DropdownItem>,
        <DropdownItem
            key='duplicate-role'
            onClick={(role) => {
                // duplicate role action
            }}
        >
            {t`Duplicate role`}
        </DropdownItem>,
        <DropdownItem
            key='delete-role'
            onClick={(role) => {
            // delete role action
            }}
        >
          {t`Delete role`}
        </DropdownItem>,
    ]}
    params={params}
    updateParams={(p) => {
        this.updateParams(p, () => this.queryGroup());
    }}
    isUserManagementDisabled={isUserMgmtDisabled}
/>
 ```
### Expanded table 
![Screenshot from 2022-03-14 15-16-02](https://user-images.githubusercontent.com/19647757/158201323-ecf1bcd2-5200-4d8f-b6cd-eebea5c7b108.png)
### Dropdown actions
![Screenshot from 2022-03-14 15-16-14](https://user-images.githubusercontent.com/19647757/158201332-4e5a2551-dd04-4cd7-ab81-2dd39e977e65.png)
### Dropdown actions 2
![Screenshot from 2022-03-14 15-19-08](https://user-images.githubusercontent.com/19647757/158201344-58eb065a-40e9-4aef-a31e-e8eecf97965f.png)


 Would this work for your `Role Page`? @ShaiahWren

ShaiahWren commented on src/components/role-list-table/role-list-table.tsx:
```diff
@@ -0,0 +1,186 @@
+  const [expandedRoles, setExpandedRoles] = useState<string[]>([]);
+
+  const isRoleExpanded = (role) => expandedRoles.includes(role);
+
```
`const [isExampleCompact, setIsExampleCompact] = React.useState(true);`

ShaiahWren replied:
Shall we add this piece of state to have the option of a compact table? Tina and Taufique asked if this was possible in one of our team meetings about the refactor.

ShaiahWren commented on src/components/role-list-table/role-list-table.tsx:
```diff
@@ -0,0 +1,186 @@
+  };
+
+  return (
+    <TableComposable aria-label='role-list-table' data-cy='RoleListTable'>
```
```suggestion
    <TableComposable variant={isRoleCompact ? 'compact' : undefined} aria-label='role-list-table' data-cy='RoleListTable' >
```

ShaiahWren commented on src/components/role-list-table/role-list-table.tsx:
```diff
@@ -0,0 +1,186 @@
+            <Td>{name}</Td>
+            <Td>{description}</Td>
+            {dropdownActions && (
+              <Td style={{ paddingRight: '0px', textAlign: 'right' }}>
```
This styling will be removed with the ListItemActions component on https://github.com/ansible/ansible-hub-ui/pull/1737, but no need to worry about that now :)

jerabekjiri commented:
#### Added new prop `variant` (`expandable`, `radio`, `checkbox`) 
- Checkbox variant
![Screenshot from 2022-03-16 22-18-35](https://user-images.githubusercontent.com/19647757/158696114-0d4442a0-9bea-4718-9421-f1b5891ddafb.png)

- Radio variant
![Screenshot from 2022-03-16 22-19-12](https://user-images.githubusercontent.com/19647757/158696155-8b00bd3f-94c1-4bb6-9fc3-bf7dfb0116e4.png)

jerabekjiri commented:
@ShaiahWren Looking at [AAH-1129](https://issues.redhat.com/browse/AAH-1129), I think we will need both radio and checkboxes. Radio buttons for selecting a role and checkbox for adding roles to the group.
![HUB-RBAC-Namespaces-Group List View-Wizard Step 1-Selected](https://user-images.githubusercontent.com/19647757/159005633-0f604bbd-1da3-4f24-99a1-496efae8763d.png)
 
![HUB-RBAC-Add Wizard-Step 1-Filter](https://user-images.githubusercontent.com/19647757/159005682-7e9daa54-6a89-4fb6-bf60-e1b40e55a653.png)

jerabekjiri commented:
- Adding `isStickyHeader` prop, which could be useful in longer list for example in `WizardModal` 
![Screenshot from 2022-03-21 12-54-19](https://user-images.githubusercontent.com/19647757/159256121-daad0570-1d6d-421d-9911-c84f70c495cf.png)

himdel commented:
> I decided to use a functional component and use hooks (useState). I've seen it's used for collection signing (@himdel is that okay? I can revert it to class component).

It's OK, also discussed with David, if we're comfortable using functional components for new stuff, nothing is stopping us :)

---

Otherwise LGTM :+1:, not seeing any issues,

perhaps just one confusion .. I'm not sure the expandable role table is the same table we'd be using in the wizard?
(I'm pretty sure it is the one we'd use on the detail screen, but I think the table of groups with radios, and roles with checkboxes is a different table from the one here.)

jerabekjiri commented:
### Reworked the RoleListTable
moved the logic from the component and added `ExpandableRow`, `CheckboxRow` and `RadioRow`, which can be used instead of the `variant` prop.

- Radio variant
```tsx
<RoleListTable
  isStickyHeader={true}
  params={this.state.params}
  updateParams={this.updateParams}
  isCompact={true}
>
  {roles.map((role, i) => (
    <RadioRow
      key={role.name}
      isSelected={this.state.selectedRole === role.name}
      onSelect={() =>
        this.setState({
          selectedRole: role.name,
        })
      }
    >
      <td>{role.name}</td>
      <td>{role.description}</td>
    </RadioRow>
  ))}
</RoleListTable>
```
- Checkbox variant
```tsx
<RoleListTable
  isStickyHeader={true}
  params={this.state.params}
  updateParams={this.updateParams}
  isCompact={true}
>
  {roles.map((role, i) => (
    <RadioRow
      key={role.name}
      isSelected={this.state.selectedRoles.includes(role.name)}
      onSelect={() =>
        this.setState({
          selectedRoles: [...this.state.selectedRoles, role.name],
        })
      }
    >
      <td>{role.name}</td>
      <td>{role.description}</td>
    </RadioRow>
  ))}
</RoleListTable>
```

- and expandable variant with permissions chips
```tsx
<RoleListTable
  isStickyHeader={true}
  params={this.state.params}
  updateParams={this.updateParams}
  isCompact={true}
>
  {roles.map(role => (
    <ExpandableRow
      key={role.name}
      expandableRowContent={
        <>
          {groups.map((group) => (
            <Flex
              style={{ marginTop: '16px' }}
              alignItems={{ default: 'alignItemsCenter' }}
              key={group.name}
              className={group.name}
            >
              <FlexItem style={{ minWidth: '200px' }}>
                {i18n._(group.label)}
              </FlexItem>
              <FlexItem grow={{ default: 'grow' }}>
                <PermissionChipSelector
                  availablePermissions={group.object_permissions
                    .filter(
                      (perm) =>
                        !selectedPermissions.find(
                          (selected) => selected === perm,
                        ),
                    )
                    .map((value) =>
                      twoWayMapper(value, filteredPermissions),
                    )
                    .sort()}
                  selectedPermissions={selectedPermissions
                    .filter((selected) =>
                      group.object_permissions.find(
                        (perm) => selected === perm,
                      ),
                    )
                    .map((value) =>
                      twoWayMapper(value, filteredPermissions),
                    )}
                  menuAppendTo='inline'
                  multilingual={true}
                  isViewOnly={true}
                />
              </FlexItem>
            </Flex>
          ))}
        </>
      }
    >
      <td>{role.name}</td>
      <td>{role.description}</td>
      <ListItemActions kebabItems={dropdownItems} buttons={buttons} />
    </ExpandableRow>
  }))
</RoleListTable>
```
@himdel @ShaiahWren would this work better?

himdel commented on src/components/permissions/permission-chip-selector.tsx:
```diff
@@ -13,7 +13,7 @@ import {
   availablePermissions: string[];
   selectedPermissions: string[];
-  setSelected: (selected: string[]) => void;
+  setSelected?: (selected: string[]) => void;
```
is this bit intentional? (doesn't *seem* related?)

jerabekjiri replied:
yes, so I can use this component as readonly component

himdel reviewed (approved):
LGTM, nice separation :)

jerabekjiri commented:
Merging, so this can unblock @ShaiahWren
